AirCasting/ABConnector/SDCardAirBeamServices.swift

```diff
@@ -153,8 +160,33 @@ class BluetoothSDCardAirBeamServices: SDCardAirBeamServices {
     private func finishSync(completion: () -> Void) {
         self.bluetoothManager.unsubscribeCharacteristicObserver(self.dataCharacteristicObserver!)
         self.bluetoothManager.unsubscribeCharacteristicObserver(self.metadataCharacteristicObserver!)
+        expectedMeasurementsCount = [:]
+        receivedMeasurementsCount = [:]
+        if let timerToken = timerToken {
+            timer.stop(token: timerToken)
+        }
         completion()
     }
+    
+    private func startMonitoringForEnd(completion: @escaping () -> Void) -> Void {
+        var checkedMeasurementsCount: [SDCardSessionType: Int] = [:]
+        timerToken = timer.schedule(every: 1) { [weak self] in
+            guard let self = self else { return }
+            Log.debug("Checking with:\n expected \(self.expectedMeasurementsCount)\n received \(self.receivedMeasurementsCount)\n checked \(checkedMeasurementsCount)")
+            guard checkedMeasurementsCount != self.receivedMeasurementsCount else {
+                Log.debug("NO NEW MEASUREMENT IN 1 SEC")
+                completion()
+                return
+            }
+            checkedMeasurementsCount = self.receivedMeasurementsCount
+            for temp in self.receivedMeasurementsCount {
+                if self.expectedMeasurementsCount[temp.key] ?? 0 < temp.value {
```

Comment:
As far as I cal tell this is checked on a different thread than being written too (checking here on the main thread, writing on a arbitrary BT manager queue). Potential race condition. Please use locks or queues to synchronize access